cpp/src/io/text/multibyte_split.cu

```diff
@@ -110,11 +141,32 @@ struct PatternScan {
 // it begins in. From there, each thread can then take deterministic action. In this case, the
 // deterministic action is counting and outputting delimiter offsets when a delimiter is found.
 
+struct singlepass_offset {
+  // magnitude stores the offset, sign bit stores whether we are past the end of the last delimiter
+  int64_t value;
+
+  constexpr void init(bool is_delim, bool is_past_range)
+  {
+    value = is_past_range ? -is_delim : is_delim;
+  }
+
+  [[nodiscard]] constexpr int64_t offset() const { return value < 0 ? -value : value; }
+
+  [[nodiscard]] constexpr bool is_past_end() { return value < 0; }
+
+  friend constexpr singlepass_offset operator+(singlepass_offset lhs, singlepass_offset rhs)
```

Comment:
this is a [hidden friend](https://www.justsoftwaresolutions.co.uk/cplusplus/hidden-friends.html) declaration that will not be found if the parameters need an implicit conversion to fit the signature. It's not really necessary, but I like the idiom to avoid polluting the global namespace.

Reply:
That's an interesting thing. I didn't know that.